## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [2]:
import pandas as pd

In [5]:
data = pd.read_csv('horse_data.csv',header=None,na_values = '?' )
data.columns = ['surgery','age','Hospital Number','rectal temperature','pulse','respiratory rate',
                'temperature of extremities','peripheral pulse','mucous membranes','capillary refill time',
               'pain','peristalsis','abdominal_distension','nasogastric tube','nasogastric reflux',
               'nasogastric reflux PH','rectal examination','abdomen','packed cell volume','total protein',
               'abdominocentesis appearance','abdomcentesis total protein','outcome','surgical lesion',
               'type of lesion','type of lesion1','type of lesion2','cp_data']
data

surgery  age  Hospital Number  rectal temperature  pulse  \
0        2.0    1           530101                38.5   66.0   
1        1.0    1           534817                39.2   88.0   
2        2.0    1           530334                38.3   40.0   
3        1.0    9          5290409                39.1  164.0   
4        2.0    1           530255                37.3  104.0   
..       ...  ...              ...                 ...    ...   
295      1.0    1           533886                 NaN  120.0   
296      2.0    1           527702                37.2   72.0   
297      1.0    1           529386                37.5   72.0   
298      1.0    1           530612                36.5  100.0   
299      1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
0                28.0                         3.0               3.0   
1                20.0                         NaN               NaN   
2                24.0                         1.0               1.0   
3                84.0                         4.0               1.0   
4                35.0                         NaN               NaN   
..                ...                         ...               ...   
295              70.0                         4.0               NaN   
296              24.0                         3.0               2.0   
297              30.0                         4.0               3.0   
298              24.0                         3.0               3.0   
299              20.0                         NaN               NaN   

     mucous membranes  capillary refill time  ...  packed cell volume  \
0                 NaN                    2.0  ...                45.0   
1                 4.0                    1.0  ...                50.0   
2                 3.0                    1.0  ...                33.0   
3                 6.0                    2.0  ...                48.0   
4                 6.0                    2.0  ...                74.0   
..                ...                    ...  ...                 ...   
295               4.0                    2.0  ...                55.0   
296               4.0                    2.0  ...                44.0   
297               4.0                    1.0  ...                60.0   
298               3.0                    1.0  ...                50.0   
299               NaN                    NaN  ...                36.0   

     total protein  abdominocentesis appearance  abdomcentesis total protein  \
0              8.4                          NaN                          NaN   
1             85.0                          2.0                          2.0   
2              6.7                          NaN                          NaN   
3              7.2                          3.0                          5.3   
4              7.4                          NaN                          NaN   
..             ...                          ...                          ...   
295           65.0                          NaN                          NaN   
296            NaN                          3.0                          3.3   
297            6.8                          NaN                          NaN   
298            6.0                          3.0                          3.4   
299           62.0                          1.0                          1.0   

     outcome  surgical lesion  type of lesion  type of lesion1  \
0        2.0                2           11300                0   
1        3.0                2            2208                0   
2        1.0                2               0                0   
3        2.0                1            2208                0   
4        2.0                2            4300                0   
..       ...              ...             ...              ...   
295      3.0                2            3205                0   
296  

In [6]:
horse = data[['surgery','age','pulse','abdominal_distension','abdomen','outcome',
              'surgical lesion','type of lesion']] # выбираю 8 столбцов
horse

surgery  age  pulse  abdominal_distension  abdomen  outcome  \
0        2.0    1   66.0                   4.0      5.0      2.0   
1        1.0    1   88.0                   2.0      2.0      3.0   
2        2.0    1   40.0                   1.0      1.0      1.0   
3        1.0    9  164.0                   4.0      NaN      2.0   
4        2.0    1  104.0                   NaN      NaN      2.0   
..       ...  ...    ...                   ...      ...      ...   
295      1.0    1  120.0                   NaN      5.0      3.0   
296      2.0    1   72.0                   3.0      4.0      3.0   
297      1.0    1   72.0                   3.0      5.0      2.0   
298      1.0    1  100.0                   3.0      4.0      1.0   
299      1.0    1   40.0                   NaN      1.0      3.0   

     surgical lesion  type of lesion  
0                  2           11300  
1                  2            2208  
2                  2               0  
3                  1            2208  
4                  2            4300  
..               ...             ...  
295                2            3205  
296                1            2208  
297                1            3205  
298                1            2208  
299                2            6112  

[300 rows x 8 columns]

In [4]:
horse.describe()

surgery         age       pulse  abdominal distension     abdomen  \
count  299.000000  300.000000  276.000000            244.000000  182.000000   
mean     1.397993    1.640000   71.913043              2.266393    3.692308   
std      0.490305    2.173972   28.630557              1.065131    1.491725   
min      1.000000    1.000000   30.000000              1.000000    1.000000   
25%      1.000000    1.000000   48.000000              1.000000    2.000000   
50%      1.000000    1.000000   64.000000              2.000000    4.000000   
75%      2.000000    1.000000   88.000000              3.000000    5.000000   
max      2.000000    9.000000  184.000000              4.000000    5.000000   

          outcome  surgical lesion  type of lesion  
count  299.000000       300.000000      300.000000  
mean     1.551839         1.363333     3657.880000  
std      0.737187         0.481763     5399.513513  
min      1.000000         1.000000        0.000000  
25%      1.000000         1.000000     2111.750000  
50%      1.000000         1.000000     2673.500000  
75%      2.000000         2.000000     3209.000000  
max      3.000000         2.000000    41110.000000

In [22]:
# Pulse
print(horse.pulse.max() - horse.pulse.min(), 'размах')
print(horse.pulse.mode()[0],'мода') 
print(horse.pulse.median(),'медиана')
print(horse.pulse.var(),'дисперсия')

154.0 размах
48.0 мода
64.0 медиана
819.7087747035575 дисперсия


48.0 мода -->преобладают животные с болезненными поражениями или страдающие от кровеносного шока

In [21]:
horse[(horse['pulse'] > 40.0)]#лошади с завышенным пульсом

surgery  age  pulse  abdominal distension  abdomen  outcome  \
0        2.0    1   66.0                   4.0      5.0      2.0   
1        1.0    1   88.0                   2.0      2.0      3.0   
3        1.0    9  164.0                   4.0      NaN      2.0   
4        2.0    1  104.0                   NaN      NaN      2.0   
6        1.0    1   48.0                   3.0      5.0      1.0   
..       ...  ...    ...                   ...      ...      ...   
293      1.0    1   78.0                   NaN      4.0      3.0   
295      1.0    1  120.0                   NaN      5.0      3.0   
296      2.0    1   72.0                   3.0      4.0      3.0   
297      1.0    1   72.0                   3.0      5.0      2.0   
298      1.0    1  100.0                   3.0      4.0      1.0   

     surgical lesion  type of lesion  
0                  2           11300  
1                  2            2208  
3                  1            2208  
4                  2            4300  
6                  1            3124  
..               ...             ...  
293                2            2209  
295                2            3205  
296                1            2208  
297                1            3205  
298                1            2208  

[252 rows x 8 columns]

In [7]:
print(horse.age.mode()[0])
#наиболее часто встречаются взрослые лошади(возраст более 6 месяцев)

1


## краткий вывод:
наиболее часто встречаются взрослые лошади, у которых 
#### 1) преобладают болезненные поражения или страдающие от кровеносного шока 
#### 2) завышенный пульс -252 
#### 3) лечились с помощью хирургического вмешательства -120

# Задание 2. Работа с выбросами¶

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.


In [120]:
# нормальный пульс 30-40, предполагаю что пульс больше чем в два раза от нормального состояния является выбросом
badpulse = horse[horse.pulse < 80]
badpulse

surgery  age  pulse  abdominal distension  abdomen  outcome  \
0        2.0    1   66.0                   4.0      5.0      2.0   
2        2.0    1   40.0                   1.0      1.0      1.0   
6        1.0    1   48.0                   3.0      5.0      1.0   
7        1.0    1   60.0                   2.0      4.0      2.0   
10       1.0    1   66.0                   1.0      5.0      1.0   
..       ...  ...    ...                   ...      ...      ...   
293      1.0    1   78.0                   NaN      4.0      3.0   
294      2.0    1   40.0                   1.0      2.0      1.0   
296      2.0    1   72.0                   3.0      4.0      3.0   
297      1.0    1   72.0                   3.0      5.0      2.0   
299      1.0    1   40.0                   NaN      1.0      3.0   

     surgical lesion  type of lesion  
0                  2           11300  
2                  2               0  
6                  1            3124  
7                  1            2208  
10                 1            2124  
..               ...             ...  
293                2            2209  
294                2               0  
296                1            2208  
297                1            3205  
299                2            6112  

[177 rows x 8 columns]

In [127]:
# Pulse до выброса
print(horse.pulse.max())
print(horse.pulse.min())
print(horse.pulse.max() - horse.pulse.min(), 'размах')
print(horse.pulse.mode()[0],'мода') 
print(horse.pulse.median(),'медиана')
print(horse.pulse.var(),'дисперсия')

184.0
30.0
154.0 размах
48.0 мода
64.0 медиана
819.7087747035575 дисперсия


In [129]:
# pulse после выброса
print(badpulse.pulse.max())
print(badpulse.pulse.min())
print(badpulse.pulse.max() - badpulse.pulse.min(), 'размах')
print(badpulse.pulse.mode()[0],'мода') 
print(badpulse.pulse.median(),'медиана')
print(badpulse.pulse.var(),'дисперсия')

78.0
30.0
48.0 размах
48.0 мода
52.0 медиана
126.34501797637395 дисперсия


In [22]:
#выбросы через межквантильный размах
q1 = horse.pulse.quantile(0.25)
q3 = horse.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliera = horse[horse.pulse.between(lower_bound,upper_bound,inclusive=True)]
remove_outliera

surgery  age  pulse  abdominal distension  abdomen  outcome  \
0        2.0    1   66.0                   4.0      5.0      2.0   
1        1.0    1   88.0                   2.0      2.0      3.0   
2        2.0    1   40.0                   1.0      1.0      1.0   
4        2.0    1  104.0                   NaN      NaN      2.0   
6        1.0    1   48.0                   3.0      5.0      1.0   
..       ...  ...    ...                   ...      ...      ...   
295      1.0    1  120.0                   NaN      5.0      3.0   
296      2.0    1   72.0                   3.0      4.0      3.0   
297      1.0    1   72.0                   3.0      5.0      2.0   
298      1.0    1  100.0                   3.0      4.0      1.0   
299      1.0    1   40.0                   NaN      1.0      3.0   

     surgical lesion  type of lesion  
0                  2           11300  
1                  2            2208  
2                  2               0  
4                  2            4300  
6                  1            3124  
..               ...             ...  
295                2            3205  
296                1            2208  
297                1            3205  
298                1            2208  
299                2            6112  

[271 rows x 8 columns]

In [128]:
pd.concat([horse,remove_outliera]).drop_duplicates(keep=False)# выбросы

surgery  age  pulse  abdominal distension  abdomen  outcome  \
3        1.0    9  164.0                   4.0      NaN      2.0   
28       1.0    1    NaN                   NaN      NaN      2.0   
41       2.0    9  150.0                   NaN      NaN      1.0   
52       2.0    1    NaN                   1.0      2.0      1.0   
55       1.0    9  160.0                   4.0      NaN      2.0   
56       1.0    1    NaN                   NaN      NaN      1.0   
58       1.0    1    NaN                   3.0      4.0      2.0   
74       1.0    9    NaN                   NaN      NaN      2.0   
83       1.0    1    NaN                   4.0      NaN      2.0   
93       2.0    1    NaN                   3.0      5.0      2.0   
115      2.0    1    NaN                   2.0      NaN      1.0   
117      2.0    1    NaN                   3.0      5.0      3.0   
126      1.0    1    NaN                   2.0      NaN      1.0   
151      1.0    1    NaN                   2.0      4.0      2.0   
160      1.0    1    NaN                   NaN      5.0      1.0   
174      2.0    1    NaN                   4.0      NaN      2.0   
175      1.0    1    NaN                   NaN      NaN      1.0   
182      1.0    1    NaN                   4.0      NaN      2.0   
204      1.0    1    NaN                   2.0      NaN      1.0   
207      1.0    1    NaN                   NaN      NaN      2.0   
216      2.0    1    NaN                   NaN      NaN      1.0   
227      1.0    9    NaN                   NaN      NaN      1.0   
255      1.0    9  184.0                   3.0      NaN      2.0   
275      1.0    9  150.0                   2.0      NaN      2.0   
288      1.0    1    NaN                   NaN      NaN      1.0   

     surgical lesion  type of lesion  
3                  1            2208  
28                 2            4300  
41                 1            9400  
52                 2            3111  
55                 1            7111  
56                 1            3112  
58                 1            3205  
74                 1           11124  
83                 1            3205  
93                 1            3205  
115                2               0  
117                1            4205  
126                1           31110  
151                1            2207  
160                1            4205  
174                1            3400  
175                1            2208  
182                1            2205  
204                1            3115  
207                1            2208  
216                2               0  
227                1            9000  
255                1            7111  
275                1            4207  
288                1            2124

## вывод:
работая с выбросами по столбцу "Pulse" я увидел,что большинство данных отсутсвуют и не попадают в указанный диапозон.
Предполагаю,что причина выбросов является не качественный эксперимент, т.е. экспериментальные ошибки и ошибка ввода данных.Слишком велико значение размаха, из-за этого данные искажаются.


# Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [7]:
for col in horse.columns:
    pot_missing = horse[col].isnull().mean()
    print(f'{col} - {pot_missing:.1%}')

surgery - 0.3%
age - 0.0%
pulse - 8.0%
abdominal_distension - 18.7%
abdomen - 39.3%
outcome - 0.3%
surgical lesion - 0.0%
type of lesion - 0.0%


In [25]:
horse['age'].fillna('no info').isna().sum()

0

In [35]:
horse2 = data[['surgery','age','pulse','abdominal_distension','abdomen','outcome',
              'surgical lesion','type of lesion']]

In [49]:
#заполнение пропусков в столбце abdomen на значение mode(использую моду как часто используемое значение в столбце)
horse2.abdomen.fillna(horse.abdomen.median() ,inplace=True)

In [48]:
#заполнение пропусков в столбце abdominal distension на значение mode
horse2.abdominal_distension.fillna(horse.abdominal_distension.median() ,inplace=True)

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


#### заполнение пропусков с помощью группировки(использую медиану для заполнения,так как результат зависел от хирургического вмешательства и от значения пульса)


In [46]:
horse2.pulse.fillna(horse2.groupby(['outcome','surgery']).pulse.transform('median'),inplace=True)
horse2.groupby(['outcome','surgery']).pulse.median()

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


outcome  surgery
1.0      1.0        60.0
         2.0        52.0
2.0      1.0        84.0
         2.0        92.0
3.0      1.0        80.0
         2.0        85.0
Name: pulse, dtype: float64

In [52]:
# датафрейм без пропусков
horse2

surgery  age  pulse  abdominal_distension  abdomen  outcome  \
0        2.0    1   66.0                   4.0      5.0      2.0   
1        1.0    1   88.0                   2.0      2.0      3.0   
2        2.0    1   40.0                   1.0      1.0      1.0   
3        1.0    9  164.0                   4.0      4.0      2.0   
4        2.0    1  104.0                   2.0      4.0      2.0   
..       ...  ...    ...                   ...      ...      ...   
295      1.0    1  120.0                   2.0      5.0      3.0   
296      2.0    1   72.0                   3.0      4.0      3.0   
297      1.0    1   72.0                   3.0      5.0      2.0   
298      1.0    1  100.0                   3.0      4.0      1.0   
299      1.0    1   40.0                   2.0      1.0      3.0   

     surgical lesion  type of lesion  
0                  2           11300  
1                  2            2208  
2                  2               0  
3                  1            2208  
4                  2            4300  
..               ...             ...  
295                2            3205  
296                1            2208  
297                1            3205  
298                1            2208  
299                2            6112  

[300 rows x 8 columns]